File: src/my_project/interviews_typescript/top_150_questions_round_1/ex_113_find_first_and_last_position_of_element.ts
function searchRange(nums: number[], target: number): number[] {
    if (nums.length === 0) {
        return [-1, -1];
    }
    
    // Find the leftmost (first) occurrence
    const left = findLeft(nums, target);
    if (left === -1) {
        return [-1, -1];
    }
    
    // Find the rightmost (last) occurrence
    const right = findRight(nums, target);
    
    return [left, right];
}

function findLeft(nums: number[], target: number): number {
    let left = 0;
    let right = nums.length - 1;
    let result = -1;
    
    while (left <= right) {
        const mid = Math.floor(left + (right - left) / 2);
        
        if (nums[mid] === target) {
            result = mid;
            right = mid - 1; // Continue searching in the left half
        } else if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    
    return result;
}

function findRight(nums: number[], target: number): number {
    let left = 0;
    let right = nums.length - 1;
    let result = -1;
    
    while (left <= right) {
        const mid = Math.floor(left + (right - left) / 2);
        
        if (nums[mid] === target) {
            result = mid;
            left = mid + 1; // Continue searching in the right half
        } else if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    
    return result;
}

File: src/my_project/interviews/top_150_questions_round_22/ex_113_find_first_and_last_position_of_element.py
from typing import List


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        """
        Find the starting and ending position of a target value in a sorted array.
        Time Complexity: O(log n) using binary search
        Space Complexity: O(1)
        """
        if not nums:
            return [-1, -1]
        
        # Find the leftmost (first) occurrence
        left = self.findLeft(nums, target)
        if left == -1:
            return [-1, -1]
        
        # Find the rightmost (last) occurrence
        right = self.findRight(nums, target)
        
        return [left, right]
    
    def findLeft(self, nums: List[int], target: int) -> int:
        """Find the leftmost occurrence of target"""
        left, right = 0, len(nums) - 1
        result = -1
        
        while left <= right:
            mid = left + (right - left) // 2
            
            if nums[mid] == target:
                result = mid
                right = mid - 1  # Continue searching in the left half
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        
        return result
    
    def findRight(self, nums: List[int], target: int) -> int:
        """Find the rightmost occurrence of target"""
        left, right = 0, len(nums) - 1
        result = -1
        
        while left <= right:
            mid = left + (right - left) // 2
            
            if nums[mid] == target:
                result = mid
                left = mid + 1  # Continue searching in the right half
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        
        return result
        